File: server/src/routes/b2b/me.py
from flask import Blueprint, current_app, jsonify, g, request
from src.routes.auth.middleware import login_required
from .middleware import tenant_required
from src.types.tenant import Tenant
import hmac
import hashlib

bp = Blueprint("me", __name__, url_prefix="/me")


@bp.route("/", methods=["GET"])
@login_required
@tenant_required
def get_me():
    user: Tenant = g.tenant
    user["_id"] = str(user["_id"])
    return jsonify(user)


@bp.route("/account", methods=["PATCH"])
@login_required
@tenant_required
def update_account():
    user: Tenant = g.tenant
    account = request.json
    current_app.db.tenants.update_one({"_id": user["_id"]}, {"$set": account})
    return '', 204


@bp.route("/webhook", methods=["PATCH"])
@login_required
@tenant_required
def update_webhook():
    user: Tenant = g.tenant
    webhook = request.json
    current_app.db.tenants.update_one(
        {"_id": user["_id"]}, {"$set": {"settings.webhookSecret": webhook["webhookSecret"], "settings.callbackUrl": webhook["callbackUrl"]}})
    return '', 204


@bp.route("/cors-origins", methods=["PATCH"])
@login_required
@tenant_required
def update_security():
    user: Tenant = g.tenant
    security = request.json
    current_app.db.tenants.update_one(
        {"_id": user["_id"]}, {"$set": {"settings.allowedOrigins": security["allowedOrigins"]}})
    return '', 204


@bp.route("/api-key", methods=["PATCH"])
@login_required
@tenant_required
def update_api_key():
    user: Tenant = g.tenant
    apiKey = request.json["apiKey"]
    print(apiKey)
    apiKeyHash = hmac.new(current_app.config["HASH_SALT"].encode(
        "utf-8"), apiKey.encode("utf-8"), hashlib.sha256).hexdigest()
    current_app.db.tenants.update_one(
        {"_id": user["_id"]}, {"$set": {"apiKeyHash": apiKeyHash}})
    return '', 204

File: b2b-client/src/api/me.ts
import type { Account, Security, Tenant, Webhook } from "@/types";

export const GetMe = async (): Promise<Tenant> => {
  const res = await fetch(`${import.meta.env.VITE_API_URL}/b2b/me/`, {
    method: "GET",
    credentials: "include",
  });

  if (!res.ok) {
    throw new Error("Failed to fetch jobs");
  }

  return await res.json();
};

export const UpdateAccount = async (account: Account): Promise<void> => {
  const res = await fetch(`${import.meta.env.VITE_API_URL}/b2b/me/account/`, {
    method: "PATCH",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(account),
  });

  if (!res.ok) {
    throw new Error("Failed to update account");
  }
};

export const UpdateWebhook = async (webhook: Webhook): Promise<void> => {
  const res = await fetch(`${import.meta.env.VITE_API_URL}/b2b/me/webhook/`, {
    method: "PATCH",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(webhook),
  });

  if (!res.ok) {
    throw new Error("Failed to update webhook");
  }
};

export const UpdateCorsOrigins = async (security: Security): Promise<void> => {
  const res = await fetch(`${import.meta.env.VITE_API_URL}/b2b/me/cors-origins/`, {
    method: "PATCH",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(security),
  });

  if (!res.ok) {
    throw new Error("Failed to update cors origins");
  }
};

export const UpdateApiKey = async (apiKey: string): Promise<void> => {
  const res = await fetch(`${import.meta.env.VITE_API_URL}/b2b/me/api-key/`, {
    method: "PATCH",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ apiKey }),
  });

  if (!res.ok) {
    throw new Error("Failed to update api key");
  }
};
